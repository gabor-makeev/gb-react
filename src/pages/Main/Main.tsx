import { FC } from 'react';
import { NewsContainer } from 'components/NewsContainer/NewsContainer';

export const Main: FC = () => {
  return (
    <>
      <NewsContainer />
    </>
  );
};
