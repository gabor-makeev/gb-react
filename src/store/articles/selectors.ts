import { StoreState } from 'src/store';

export const selectArticles = (state: StoreState) => state.articles.articles;

export const selectLoading = (state: StoreState) => state.articles.loading;

export const selectError = (state: StoreState) => state.articles.error;
