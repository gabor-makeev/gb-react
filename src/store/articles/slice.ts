import { Article } from 'src/default-types';
import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { api } from 'src/constants';

interface ArticlesState {
  articles: Article[];
  loading: boolean;
  error: string;
}

const initialState: ArticlesState = {
  articles: [],
  loading: false,
  error: '',
};

export const articlesSlice = createSlice({
  name: 'articles',
  initialState,
  reducers: {
    addArticles: (state, action: PayloadAction<Article[]>) => {
      state.articles = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchArticles.pending, (state) => {
        state.loading = true;
        state.error = '';
      })
      .addCase(fetchArticles.fulfilled, (state, action) => {
        state.loading = false;
        state.articles = action.payload as unknown as Article[];
      })
      .addCase(fetchArticles.rejected, (state, { error: { message } }) => {
        state.loading = false;
        state.error = message as string;
      });
  },
});

export const fetchArticles = createAsyncThunk<PayloadAction>(
  'articles/fetchArticles',
  async () => {
    // the line below delays the fetching of data for 1 second in order to show that 'Loading' is displayed
    await new Promise((resolve) => setTimeout(resolve, 1000));

    const res = await fetch(api);
    const data = await res.json();
    return data;
  }
);

export const articlesReducer = articlesSlice.reducer;
