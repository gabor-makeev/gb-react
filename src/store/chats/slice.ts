import { createAction, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Dispatch } from 'redux';
import { getAuth } from 'firebase/auth';
import { IClientUserChat, IFirebaseMessage } from 'src/default-types';
import { UserService } from 'src/services/firebase/Service/UserService';

interface ChatsState {
  content: IClientUserChat[];
}

const initialState: ChatsState = { content: [] };

export const chatsSlice = createSlice({
  name: 'chats',
  initialState,
  reducers: {
    setChats: (state, action: PayloadAction<IClientUserChat[]>) => {
      state.content = action.payload;
    },
  },
});

export const sendMessageWithBotReply = createAction<IFirebaseMessage>(
  'chats/sendMessageWithBotReply'
);

export const initChatsTracking = () => (dispatch: Dispatch) => {
  const user = getAuth().currentUser;

  if (user?.email) {
    return UserService.subscribeToChats(user?.email, (chats) => {
      dispatch(chatsSlice.actions.setChats(chats));
    });
  }
};

export const { setChats } = chatsSlice.actions;
export const chatsReducer = chatsSlice.reducer;
