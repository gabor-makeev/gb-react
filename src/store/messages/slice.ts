import { Message } from 'src/default-types';
import { createAction, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { nanoid } from 'nanoid';

type MessageItemWithId = { id: string } & Message;

export interface MessagesState {
  [key: string]: MessageItemWithId[];
}

const initialState: MessagesState = {};

export const messagesSlice = createSlice({
  name: 'messages',
  initialState,
  reducers: {
    addChat: (state, action: PayloadAction<{ chatName: string }>) => {
      state[action.payload.chatName] = [];
    },
    deleteChat: (state, action: PayloadAction<{ chatName: string }>) => {
      delete state[action.payload.chatName];
    },
    addMessage: (
      state,
      action: PayloadAction<{ chatName: string; message: Message }>
    ) => {
      state[action.payload.chatName].push({
        id: nanoid(),
        author: action.payload.message.author,
        text: action.payload.message.text,
      });
    },
  },
});

export const addMessageWithSaga = createAction<{
  chatName: string;
  message: Message;
}>('messages/addMessageWithSaga');

export const { addChat, deleteChat, addMessage } = messagesSlice.actions;
export const messagesReducer = messagesSlice.reducer;
