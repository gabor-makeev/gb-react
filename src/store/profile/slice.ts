import { createSlice, PayloadAction } from '@reduxjs/toolkit';

export interface ProfileState {
  user: {
    name: string;
  };
  isPublic: boolean;
}

const initialState: ProfileState = {
  user: {
    name: 'Unknown user',
  },
  isPublic: false,
};

export const profileSlice = createSlice({
  name: 'profile',
  initialState,
  reducers: {
    changeName: (state, action: PayloadAction<string>) => {
      state.user.name = action.payload;
    },
    togglePublic: (state) => {
      state.isPublic = !state.isPublic;
    },
  },
});

export const { changeName, togglePublic } = profileSlice.actions;
export const profileReducer = profileSlice.reducer;
