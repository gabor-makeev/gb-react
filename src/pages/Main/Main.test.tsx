import React from 'react';
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Main } from 'src/pages/Main/Main';
import { configureStore } from '@reduxjs/toolkit';
import { articlesReducer } from 'store/articles/slice';

describe('Articles', () => {
  const store = configureStore({
    reducer: { articles: articlesReducer },
  });

  it('should render', () => {
    render(
      <Provider store={store}>
        <Main />
      </Provider>
    );
  });

  it('should render with snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Main />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
