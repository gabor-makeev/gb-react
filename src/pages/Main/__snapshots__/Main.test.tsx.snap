// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Articles should render with snapshot 1`] = `
<DocumentFragment>
  <h2>
    Articles
  </h2>
  <button>
    Get articles
  </button>
  <p>
    Loading
  </p>
</DocumentFragment>
`;
